src/utils.ts

```diff
@@ -487,6 +490,21 @@ export function hasAtomicOperators(doc: Document | Document[]): boolean {
   }
 
   const keys = Object.keys(doc);
+  // In this case we need to throw if all the atomic operators are undefined.
+  if (options?.ignoreUndefined) {
+    let allUndefined = true;
+    for (const key of keys) {
+      // eslint-disable-next-line no-restricted-syntax
+      if (doc[key] !== undefined) {
+        allUndefined = false;
+        break;
+      }
+    }
+    if (allUndefined) {
+      throw new MongoInvalidArgumentError('All atomic operators provided have undefined values.');
+    }
```

Comment:
I didn't expect us to throw here when we originally discussed this but I see now this is more meaningful than falling through to returning false and throwing "Update document requires atomic operators" which may not be as clear given that there are atomic op keys defined, they are just defined to something that disappears as serialization.

Can we update the error message to just be a bit more explicit about "what must be done" rather than "what is"? Something like: 'Cannot perform an update operation: all provided atomic operators must have defined values.' or 'Update operation requires defined values for any atomic operators.'

Reply:
I've updated the message to "Update operations require that all atomic operators have defined values, but none were provided"

Reply:
Perfect 🚀